Objects/exceptions.c

```diff
@@ -629,6 +629,450 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
+            if (!e) {
+                goto done;
+            }
+            PyObject *rec = exceptiongroup_split_recursive(
+                e, matcher, construct_rest);
+            Py_DECREF(e);
+            if (!rec) {
+                goto done;
+            }
+            if (!PyTuple_CheckExact(rec) || PyTuple_GET_SIZE(rec) != 2) {
+                PyErr_SetString(PyExc_RuntimeError,
+                    "Internal error: invalid value");
+                Py_DECREF(rec);
+                goto done;
+            }
+            int res = 0;
+            PyObject *e_match = PyTuple_GET_ITEM(rec, 0);
+            if (e_match != Py_None) {
+                res += PyList_Append(match_list, e_match);
+            }
+            PyObject *e_rest = PyTuple_GET_ITEM(rec, 1);
+            if (e_rest != Py_None) {
+                res += PyList_Append(rest_list, e_rest);
+            }
+            Py_DECREF(rec);
+            if (res < 0) {
+                goto done;
+            }
+        }
+
+        /* construct result */
+        match_exc = exceptiongroup_subset(eg, match_list);
+        if (!match_exc) {
+            goto done;
+        }
+
+        if (construct_rest) {
+            rest_exc = exceptiongroup_subset(eg, rest_list);
+            if (!rest_exc) {
+                goto done;
+            }
+        }
+        else {
+            rest_exc = Py_NewRef(Py_None);
+        }
+        result = PyTuple_Pack(2, match_exc, rest_exc);
+    done:
+        Py_XDECREF(match_exc);
+        Py_XDECREF(rest_exc);
+        Py_XDECREF(match_list);
+        Py_XDECREF(rest_list);
+        return result;
+    }
+}
+
+static PyObject *
+BaseExceptionGroup_split(PyObject *self, PyObject *args)
+{
+    PyObject *matcher_value = NULL;
+    if (!PyArg_UnpackTuple(args, "split", 1, 1, &matcher_value)) {
+        return NULL;
+    }
+
+    struct _exceptiongroup_matcher matcher;
+    matcher.value = matcher_value;
+    if (_set_matcher_type(&matcher) == -1) {
+        return NULL;
+    }
+    return exceptiongroup_split_recursive(
+        self, &matcher, 1 /* with_construct_rest */);
+}
+
+static PyObject *
+BaseExceptionGroup_subgroup(PyObject *self, PyObject *args)
+{
+    PyObject *matcher_value = NULL;
+    if (!PyArg_UnpackTuple(args, "subgroup", 1, 1, &matcher_value)) {
+        return NULL;
+    }
+
+    struct _exceptiongroup_matcher matcher;
+    matcher.value = matcher_value;
+    if (_set_matcher_type(&matcher) == -1) {
+        return NULL;
+    }
+    PyObject *ret = exceptiongroup_split_recursive(
+        self, &matcher, 0 /* without construct_rest */);
+
+    if (!ret) {
+        return NULL;
+    }
+    if (!PyTuple_CheckExact(ret)) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "Internal error: expected a 2-tuple");
+        Py_DECREF(ret);
+        return NULL;
+    }
+    PyObject *match = Py_NewRef(PyTuple_GetItem(ret, 0));
+    Py_DECREF(ret);
+    return match;
```

Comment:
~`PyTuple_GetItem` can return `NULL`.~ Actually, using `PyTuple_GET_ITEM` would be better, since we know that `ret` has two items if `ret != NULL`.

```suggestion
    PyObject *match = PyTuple_GET_ITEM(ret, 0);
    Py_DECREF(ret);
    return Py_NewRef(match);
```

Reply:
Agree about PyTuple_GET_ITEM. Though the incref does need to come before ret is released.

Reply:
Doh, of course! My bad.